EMP-109 add touch support

oliv3r opened the pull request.

oliv3r committed:
cmake: Set the proper project language

By setting the proper language for the project, we no longer get errors
about missing C++ compiler.

Contributes to issue EMP-109

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
Dockerfile: Add initial dockerfile

By adding a dockerfile we now have a build environment to build this
package.

Contributes to issue EMP-109

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
build: Add build scripts for manual builds

Add a few scripts to make it easier to manually build.

Contributes to issue EMP-109

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
Rewrite of the ulticontroller interface tester

This is a complete rewrite of the previous interface tester. Where the
previous version only reported the resolution and rendered the primary
colors of the display, this version adds touch input control and knobs
to twiddle some parameters.

Contributes to issue EMP-109

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

oliv3r committed:
ci: Add initial ci configuration

Add initial untested CI files needed to run in gitlab-ci.

Contributes to issue EMP-109

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

raysiudak commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+FROM registry.hub.docker.com/library/debian:stable-slim
```
should be jessie

oliv3r replied:
sadly, jessie does not contain an arm + libc compiler suite, we have to get that from emdebian. Currently, this is non-functional. I've spend half a day to get it to work to no avail.
start a container as such:
```sh
docker run --rm -it debian:stable-slim /bin/sh
```
And from there, do
```sh
wget http://emdebian.org/tools/debian/emdebian-toolchain-archive.key && \
apt-key add emdebian-toolchain-archive.key && \
echo "deb http://emdebian.org/tools/debian/ jessie main" > /etc/apt/sources.list.d/crosstools.list && \
dpkg --add-architecture armhf && \
apt-get update && \
apt-get install -y crossbuild-essential-armhf
```
and watch it all fall appart :(

It seems that debian jessie has been updated to certain newer versions of packages, but emdebian, being abanoned, hasn't kept up to date.

Even manually installing the dpkg files quickly fails due to package incompatibilities.

raysiudak replied:
Ok not worth the effort

raysiudak commented on Dockerfile:
```diff
@@ -0,0 +1,22 @@
+    rm -rf /var/lib/apt/lists/
+
+ENV PKG_CONFIG_PATH="${PKG_CONFIG_PATH:-/usr/lib/arm-linux-gnueabihf/pkgconfig/}"
+ENV CROSS_COMPILE="${CROSS_COMPILE:-arm-linux-gnueabihf-}"
```
arm-none-eabi

oliv3r replied:
actually, no. arm-none-eabi is a libc-less compiler. It is mostly intended for embedded things, such as arm based microcontrollers. The kernel and u-boot are 'special' in that they do not run in userspace, they are in effect microcontroller-like pieces of software. We are using libraries and libc thus we need the full blown arm-linux-gnueabihf suite.

raysiudak commented on build.sh:
```diff
@@ -0,0 +1,37 @@
+# SPDX-License-Identifier: AGPL-3.0+
+#
+# Copyright (C) 2018 Ultimaker B.V.
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
```
software-embedded-platform@ultimaker.com

oliv3r replied:
the contact information for the package is, I am the copyright holder however.

raysiudak commented on build_for_ultimaker.sh:
```diff
@@ -0,0 +1,32 @@
+CI_REGISTRY_IMAGE="${CI_REGISTRY_IMAGE:-registry.gitlab.com/ultimaker_bv/ulticontroller_interface_tester}"
+CI_REGISTRY_IMAGE_TAG="${CI_REGISTRY_IMAGE_TAG:-latest}"
+CROSS_COMPILE="${CROSS_COMPILE:-arm-linux-gnueabihf-}"
+WORKDIR="${WORKDIR:-/build}"
```
Maybe docker workdir would be more understandable since this script is run on the host.

oliv3r replied:
:+1:

raysiudak commented on tests/buildenv.sh:
```diff
@@ -0,0 +1,21 @@
+}
+
+trap cleanup EXIT
+
```
A bit overkill maybe.. libevdev is tested by compiling and by cmake, maybe there should be a check if all files exist.

oliv3r replied:
not sure what you mean? is the cleanup overkill?

if you are referring to the actual compile test, we are not testing if evdev is errnous, we are only checking if we can compile a test aplication using libevdev to ensure that the build environment is able to do do a compile. Or is this not what you meant?

I.e. if we can generate an evdev based program, the build environment is 'ok'.

raysiudak commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: AGPL-3.0+
+#
+# Copyright (C) 2018 Ultimaker B.V.
+# Copyright (C) 2018 Olliver Schinagl <oliver@schinagl.nl>
```
software-embedded-platform@

raysiudak commented on tests/buildenv.sh:
```diff
@@ -13,9 +13,20 @@ cleanup() {
+	
+	int main(void)
+	{
+		struct libevdev *dev;
```
dev is uninitialized, assign NULL

oliv3r replied:
dev gets its initialization value from libevdev_new().

I could have written:
```c
struct libevdev *dev = libevdev_new();
```
instead, since it saves a line in the test code i actually will :+1: 
but normally, i prefer to keep my (non simple/0/NULL) inits generally at the top of the code, rather then in init.

raysiudak commented on tests/buildenv.sh:
```diff
@@ -13,9 +13,20 @@ cleanup() {
+EOT
+
+"${CROSS_COMPILE}gcc" -o "${TEST_DIR}/compiler_test.o" -c "${TEST_DIR}/compiler_test.c" -Wall -Werror "$(pkg-config --libs --cflags libevdev)"
 "${CROSS_COMPILE}objdump" -dS "${TEST_DIR}/compiler_test.o" 1> /dev/null
```
Why dump the output?

oliv3r replied:
we use the cross compiler related objdump to verify that we are generating a cross-compile binary.
for example, if for some reason the generated code is x86 (or even nothing) rather then arm, arm-linux-gnueabihf-objdump will fail as the output is not a valid arm binary.

raysiudak commented on README.md:
```diff
@@ -0,0 +1,43 @@
+# UltiController Interface Tester
+
+The UltiController Interface Tester or ucit for short, is a simple optionally 
+statically link able application that can be used to validate the touch-screen
```
link able is one word

oliv3r replied:
i guess so :) my spell-checker disagrees though :(

raysiudak commented on README.md:
```diff
@@ -0,0 +1,43 @@
+
+The UltiController Interface Tester or ucit for short, is a simple optionally 
+statically link able application that can be used to validate the touch-screen
+available on the ulti controller.
```
should the u'lti controller' also be UltiController

oliv3r replied:
:+1:  though it's probably ulti-controller? Who knows :)

raysiudak commented on README.md:
````diff
@@ -0,0 +1,43 @@
+
+If a different build environment is needed, this can be accomplished by first
+created the docker container
+```sh docker build -t "ucit:latest" .
````
I think that the actual shell command should be on a new line, right now they are not visible on github. 
This goes for all the inline shell parts.

raysiudak commented on src/ucit.c:
```diff
@@ -6,9 +6,17 @@
  *
  */
 
+#define _GNU_SOURCE
```
Go up 7 lines and change the Copyright

oliv3r replied:
:-1: but _GNU_SOURCE is to be implemented slightly different :)

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ *
+ * Return:	framerate in milliseconds
+ */
+#define FPS(__fps)	((1 * 1000) / (__fps))
```
Can this define receive 0?

oliv3r replied:
yes, if you request to render at 0 FPS (hardcoded value) it could. Since this is a compile-time check however, the compiler checks this for us:
```
ulticontroller_interface_tester/src/ucit.c:66:32: error: division by zero [-Werror=div-by-zero]
 #define FPS(__fps) ((1 * 1000) / (__fps))

ulticontroller_interface_tester/src/ucit.c:817:22: note: in expansion of macro ‘FPS’
  printf("FPS: %d\n", FPS(0));
```
So if a programmer sets this to 0 and tries to compile it, it will fail during compilation, thus basically, no this cannot ever be 0 during runtime/production.

You are right however that if new features are added to the program, and it gets turned into FPS(variable) and variable isn't checked and passed as 0, we get a division by 0.
I guess that's a YAGNI for now. ...  damnit now i've added it.

But again, now it's just to remove a magic value.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ * @g:	8-bit green component
+ * @b:	8-bit blue component
+ */
+static struct color {
```
symbol color never used

oliv3r replied:
made it anonymous until it is actually needed

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	uint8_t g;
+	uint8_t b;
+} background_colors[] = {
+	{ .r = UINT8_MAX, .g = UINT8_MAX, .b = UINT8_MAX },
```
I agree that this is maybe easier to use and less lines of code, but defining the array like this is not very descriptive and difficult to read. I prefer naming the colors and then storing the references in an array and using something like a next_color() helper.

oliv3r replied:
for bigger applications I'd agree, but this is what you are suggesting in essence.

You say 'naming the colors and then storing them in an array. This is not naming them but storing them in an array :)

doing
```c
static struct color {
	uint8_t r;
	uint8_t g;
	uint8_t b;
};

static struct color white = {  .r = UINT8_MAX, .g = UINT8_MAX, .b = UINT8_MAX }
static struct color black = {  .r =      0x00, .g =      0x00, .b =      0x00 }
static struct color red   = {  .r = UINT8_MAX, .g =      0x00, .b =      0x00 }
static struct color green = {  .r =      0x00, .g = UINT8_MAX, .b =      0x00 }
static struct color blue  = {  .r =      0x00, .g =      0x00, .b = UINT8_MAX }

static struct color background_colors[] = {
	white,
	black,
	red,
	green,
	blue
};
```
makes it not that much better (maybe a little) but doesn't compile, which would probably add even more stuff for something trivial.

I prefer to avoid a define here, but can be convinced if you think it improves readability.
```c
#define COLOR_RED {  .r = UINT8_MAX, .g =      0x00, .b =      0x00 }
```
for example, but as it is just as much code, i'll just add a comment above the array entry :)

raysiudak commented on src/ucit.c:
```diff
@@ -6,9 +6,17 @@
  */
 
+#define _GNU_SOURCE
+#include <dirent.h>
```
My editor complains about unused imports: 
limits, input, stdarg, ioctl, stat, types

oliv3r replied:
you mean unused includes, e.g. something that's included but no function from that include used?
Your IDE is wrong :p 
- [ ] [limits.h](http://pubs.opengroup.org/onlinepubs/7908799/xsh/limits.h.html) my bad, i thought UINT8_MAX came from that, but it is needed for CHAR_BIT (rather then >> 3)
- [x] [linux/input.h](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/input.h) Is needed for [EV_ABS](https://github.com/Ultimaker/ulticontroller_interface_tester/blob/EMP-109/src/ucit.c#L412) and the like. As an application we are expected to include linux/input.h not the sub-includes afaic. But more importantly [struct event](https://github.com/Ultimaker/ulticontroller_interface_tester/blob/EMP-109/src/ucit.c#L405) comes from here.
- [x] [stdarg.h]() my IDE actually says 'file not found' as it lives in /usr/lib/ these days. I've added it on purpose however. Let me get back to you on that one, but it's quite likely related to a function that takes va_args.
- [x] [ioctl.h](http://man7.org/linux/man-pages/man2/ioctl.2.html) Here your IDE is really [failing](https://github.com/Ultimaker/ulticontroller_interface_tester/blob/EMP-109/src/ucit.c#L642).
- [x] [sys/stat.h](https://linux.die.net/man/2/open) afaik it defines [(O_RDONLY](https://github.com/Ultimaker/ulticontroller_interface_tester/blob/EMP-109/src/ucit.c#L468) etc.
- [x] [sys/types.h](https://linux.die.net/man/2/open) defines size_t and [ssize_t](https://github.com/Ultimaker/ulticontroller_interface_tester/blob/EMP-109/src/ucit.c#L351).

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ *
+ * Return:	the number of elements in an array.
+ */
+#define ARRAY_SIZE(__array) (sizeof(__array) / sizeof((__array)[0]))
```
same here, 0?

oliv3r replied:
same here? this can only be 0 if __array does not exist, as anything sizeof() calls uppon is always atleast 1 byte. Even a NULL pointer. Not sure if sizeof CAN even return 0? [Nope](https://stackoverflow.com/questions/2632021/can-sizeof-return-0-zero):)

raysiudak commented on build.sh:
```diff
@@ -0,0 +1,37 @@
+	fi
+done
+
+mkdir -p "${BUILDDIR}" || true
```
build()
make_package()

raysiudak commented on build.sh:
```diff
@@ -0,0 +1,37 @@
+make
+make package
+cd "${CWD}"
+ln -sf "${BUILDDIR}/"*.deb .
```
Why are you creating a link to the file? We should think of a standard for this in all our scripts, e.g. 
.build_armhf/artifact and the consumer retrieves it from the build dir or something in the lines with your method, but then the source dir is polluted.

oliv3r replied:
yes, but that fails with the current jedi-build structure, as it copies the *.deb from the root dir :)
The final, tested artifact should always be at an expected output location. Lets discuss this further.

For now, i'll leave the symlink in place to ensure jedi-build can find the artifact.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+			b = sat_sub(background_colors[c].b, band);
+		}
+
+		buffer[i + CHAN_R] = r ^ mask[i + CHAN_R];
```
Assigning uint32_t to uint8_t

oliv3r replied:
where? buffer is uint8_t *

raysiudak replied:
I don't know :p

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ *
+ * Return:	clamped value.
+ */
+#define clamp(__val, __int) \
```
Can this function receive zero for the __int value?

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+#define DEV_FB "/dev"
+#define FB_DEV_NAME "fb"
+
+
```
remove extra line

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ * @id:			identifier/name of the framebuffer
+ * @xres:		current resolution along the X-axis of the framebuffer
+ * @yres:		current resolution along the Y-axis of the framebuffer
+ * @bpp:		current bytes per pixel of the framebuffer
```
i think this should be bits per pixel?

oliv3r replied:
nope, most our math is using bytes per pixel. For example when we loop over a line of pixels, we set the R G and B bytes consequentially.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	printf("Usage: %s [OPTION] ... [<fb_dev> [<ev_dev>]]\n"
+	       "  -e, --evdev=<event_dev>		force event device <ev_dev>t\n"
+	       "  -f, --fbdev=<fb_dev>			force framebuffer device <fb_dev>\n"
+	       "  -t, --touchsize=<X[xY]>		input size X x Y of test pattern (default %ux%u)\n"
```
X x Y? do you mean X * Y, it is a bit confusing

oliv3r replied:
it is quite common to define resolutions as 800x600 rather then 800*600. As the later looks more like a multiplication. See [wikipedia](https://en.wikipedia.org/wiki/Display_resolution) for example.

raysiudak replied:
Yes ok but here you do not use numbers as in 1024x768 so it looks weird, but ok just leave it.

oliv3r replied:
still it's Len x width; not len multiplied by with (which an asterix implies)

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+static void usage(char *argv0)
+{
+	printf("Usage: %s [OPTION] ... [<fb_dev> [<ev_dev>]]\n"
+	       "  -e, --evdev=<event_dev>		force event device <ev_dev>t\n"
```
Why do you have options and arguments for fb_ and ev_dev?

oliv3r replied:
now it tries to automatically detect both parameters. If it miss-detects these, you can force them anyway. This due to the dynamic nature of these devices.

raysiudak replied:
Yes but now it is double with argument flags -e and -f and options. But its fine with me if you want to be able pass them as part of the execution environment.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+/**
+ * struct display_info - framebuffer display information structure
+ *
+ * @fb_dev:		file descriptor to a framebuffer device
```
Text alignment is off

oliv3r replied:
it seems like i already fixed this as i don't see it locally anymore

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	       "  -f, --fbdev=<fb_dev>			force framebuffer device <fb_dev>\n"
+	       "  -t, --touchsize=<X[xY]>		input size X x Y of test pattern (default %ux%u)\n"
+	       "  -s, --fadespeed <speed>		input fadeout speed (default %u)\n"
+	       "  -b, --banding				enable banding of the background\n"
```
Text alignment is off, tabs and spaces

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ * Note that it is currently assumed that there is no virtual space beyond
+ * the xres and yres.
+ *
+ * Note that the @buffer and @mask buffer need to be the same size as the
```
I do not see any check in the code for this

oliv3r replied:
there is no, we only have one fb_len parameter, and both need to be declared the same size. I did add a comment about using fb_len for both.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+static void background_draw(uint8_t *buffer, const uint8_t *mask, struct display_info *disp, const bool banding, const bool colorize)
+{
+	static uint8_t c = 0;
+	uint32_t i;
```
declare and assign i in the for loop

oliv3r replied:
you youngsters and your modern concepts of things. Regular C actually disallows this and while there are extensions that allow mixed declarations, I personally think it's a bad idea (tm). The exception being in for loops I suppose.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ * @yres:		current resolution along the Y-axis of the framebuffer
+ * @bpp:		current bytes per pixel of the framebuffer
+ * @fb_len:		number of bytes of the current framebuffer
+ * @line_length:	the length, in bytes, of a line of the current framebuffer
```
I think fb_len and line_length should be calculated from x_res, y_res and bpp where needed with helpers

oliv3r replied:
actually, line_length is given by the kernel and is not really calculateable as it may contain extra timings such as the front-porch and back-porch and what not. A line may not be exactly xres * bpp. Luckly this is mostly an artifact from ye-ol-days of CRTS and LCD's don't need to suffer with line_lengths that are too long. fb_len can also not be calculated due to the same reason, it is the smem_len parameter from the kernel.

The code tries to take this into account as much as possible, unless I missed something of course or was accidentally lazy.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+};
+
+/**
+ * version() - prints the program version string
```
Just wondering what the policy is about the use of capitals?

oliv3r replied:
depends on the language, python has pep8 rules, c++ generally has a few styles as well, we use a mix of pep8 and something made up ourselves.

C tends to quite commonly follow K&R rules, but even here there are many variants.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	uint8_t band = UINT8_MAX;
+	uint32_t band_gap = (disp->line_length / UINT8_MAX);
+
+	for (i = 0; i < disp->fb_len; i += disp->bpp) {
```
Extract banding responsibility to function

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		}
 	}
+
+	free(backbuffer);
```
Unreachable code

oliv3r replied:
due to the while true; yes, but once the while true breaks (due to whatever reason) we are cleaning up after ourselves.

e.g. we malloc; thus we should free. otherwise we have mallocs without free's!

anyway, with the addition of the 'print ok and abort' feature this is actually reached again.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		       int32_t x, int32_t y, uint32_t xsize, uint32_t ysize)
 {
-	int i;
+	uint32_t line;
```
declare and assign directly in the for loop

raysiudak replied:
And is see you are using line and col because x and y are already taken, to be more consistent i suggest renaming 'line' to 'row' then.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
-		fb[i + 1] = 0x00;
-		fb[i + 2] = 0x00;
+	for (line = y; line < (y + ysize); line++) {
+		uint32_t col;
```
same here

oliv3r replied:
row > line anyway!

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+			if ((coord + 3) > disp->fb_len)
+				break;
+
+			mask[coord + CHAN_R] = UINT8_MAX;
```
i see the touch mask buffer is just a mapping of framebuffer pixels and touchpoint, in that case the storage chosen here is not very efficient. A touch point (pixel) is either on or off, this almost sounds binary :P, you do not need 4 bytes per pixel. On byte with UINT8_MAX would be sufficient for XORing.

oliv3r replied:
you are right in that it is not memory efficient, but it makes the code a bit easier, as we can just have the mask buffer be the same size and iterate over it identically and saves a math operation (at the cost of more memory). Do note that while yes, it's almost a boolean, it also is a mask on the buffer, so in that regard it does make sense :)

e.g. it would have turned into
```c
buffer[i + CHAN_R] = r ^ mask[i / disp->bpp];
```
and since this is a render pipeline, performance > memory efficiency (especially for such a simple application).

I did give it a quick where, but for some reason it didn't work; so if you find it important enough I can investigate it more deeply as to why it didn't work. See here my attempt
```patch
From 9b34b20057e76c63cb88d1310f03b46ca8177e87 Mon Sep 17 00:00:00 2001
From: Olliver Schinagl <oliver@schinagl.nl>
Date: Mon, 6 Aug 2018 11:56:44 +0200
Subject: [PATCH] fixup! Rewrite of the ulticontroller interface tester

improve memory and readability at the cost of performance :(
---
 src/ucit.c | 21 ++++++++++-----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/src/ucit.c b/src/ucit.c
index 1bbcf78..c9795be 100644
--- a/src/ucit.c
+++ b/src/ucit.c
@@ -283,6 +283,7 @@ static void background_draw(uint8_t *buffer, const uint8_t *mask, struct display
 	uint32_t i;
 
 	for (i = 0; i < disp->fb_len; i += disp->bpp) {
+		uint32_t mask_i = i / disp->bpp;
 		uint8_t r = background_colors[c].r;
 		uint8_t g = background_colors[c].g;
 		uint8_t b = background_colors[c].b;
@@ -290,9 +291,9 @@ static void background_draw(uint8_t *buffer, const uint8_t *mask, struct display
 		if (banding)
 			band_pixel(disp->line_length, i, c, &r, &g, &b);
 
-		buffer[i + CHAN_R] = r ^ mask[i + CHAN_R];
-		buffer[i + CHAN_G] = g ^ mask[i + CHAN_G];
-		buffer[i + CHAN_B] = b ^ mask[i + CHAN_B];
+		buffer[i + CHAN_R] = r ^ mask[mask_i];
+		buffer[i + CHAN_G] = g ^ mask[mask_i];
+		buffer[i + CHAN_B] = b ^ mask[mask_i];
 		buffer[i + CHAN_A] = 0x00;
 	}
 
@@ -361,14 +362,11 @@ static void input_mark(uint8_t *mask, bool *matrix, const struct display_info *d
 		for (col = x; col < (x + xsize); col++) {
 			uint32_t coord;
 
-			coord = (col * disp->bpp) + (row * disp->line_length);
-			if ((coord + 3) > disp->fb_len)
+			coord = col + (row * disp->line_length);
+			if ((coord) > disp->fb_len)
 				break;
 
-			mask[coord + CHAN_R] = UINT8_MAX;
-			mask[coord + CHAN_G] = UINT8_MAX;
-			mask[coord + CHAN_B] = UINT8_MAX;
-			mask[coord + CHAN_A] = 0x00;
+			mask[coord] = UINT8_MAX;
 		}
 	}
 }
@@ -420,6 +418,7 @@ static int renderloop(struct libevdev *evdev, struct display_info *disp,
 	bool matrix[matrix_size];
 	clock_t offset;
 	uint32_t elapsed = 0;
+	uint32_t touchmask_size = disp->fb_len / disp->bpp;
 	uint8_t *backbuffer, *touchmask;
 
 	memset(disp->fb, 0x00, disp->fb_len);
@@ -430,7 +429,7 @@ static int renderloop(struct libevdev *evdev, struct display_info *disp,
 	if (!backbuffer)
 		return -ENOMEM;
 
-	touchmask = (uint8_t *)calloc(disp->fb_len, sizeof(uint8_t));
+	touchmask = (uint8_t *)calloc(touchmask_size, sizeof(uint8_t));
 	if (!touchmask)
 		return -ENOMEM;
 
@@ -460,7 +459,7 @@ static int renderloop(struct libevdev *evdev, struct display_info *disp,
 				memcpy(disp->fb, backbuffer, disp->fb_len);
 				frame_drawn = true;
 
-				input_fade(touchmask, disp->fb_len, fade);
+				input_fade(touchmask, touchmask_size, fade);
 
 				background_draw(backbuffer, touchmask, disp, banding, bg_cycle_color);
 
-- 
2.18.0
```

raysiudak replied:
It is one math operation for the index, i guess yes this would indeed disrupt pipelining ( isn't the compiler smart enough to fix this ?). 
Anyway, no it is not important enough.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ * input_fade() - helper function to fade the input events away
+ *
+ * @mask:	mask buffer to render input events into
+ * @fb_len:	size of mask buffer
```
Documentation does not match the actual argument naming

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	clock_t offset;
+	uint8_t *backbuffer, *touchmask;
+	uint32_t elapsed = 0;
+	bool frame_drawn = false, update_input = false;
```
I don't like multiple variable declarations on one line.

oliv3r replied:
initially they where more closely related; but that's not the case anymore so you are right. note that they are still related in that they are still render loop variables, one being the output, the other being the input

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		     uint32_t xsize, uint32_t ysize, uint32_t fade, const bool banding)
+{
+	clock_t offset;
+	uint8_t *backbuffer, *touchmask;
```
In general, is it not good practice to assign NULL or 0 to your pointer if you do not use it right away?

oliv3r replied:
yes and maybe.

This is just the declaration of the variable, it IS being initialized right away with the malloc at the start of the function.

In this case, the compiler will just optimize them away I would expect as the assignment is never used and overwritten straight away.

I think MISRA says everything needs to be initialized at declaration, used or unused.
However this is moot these days, as the compiler will warn you about even potentially uninitialized variable usage.

```sh
/workdir/src/ucit.c:479:13: error: 'elapsed' may be used uninitialized in this function [-Werror=maybe-uninitialized]
      elapsed++;
      ~~~~~~~^~
/workdir/src/ucit.c:429:11: note: 'elapsed' was declared here
  uint32_t elapsed;
```
so then the argument slightly different and more as to where do you initialize them by default.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	memset(disp->fb, 0x00, disp->fb_len);
+
+	backbuffer = (uint8_t *)calloc(disp->fb_len, sizeof(uint8_t));
+	if (!backbuffer)
```
if (backbuffer != NULL) or if (backbuffer != 0)  is more descriptive

oliv3r replied:
using !ptr is also VERY common in pointer land and ptr != 0 is actually wrong (while technically works if NULL = 0).
!ptr tends to be read as 'if not a valid pointer. I'm so used it now, that reading ptr != NULL takes a minute to think 'what does that mean again'. So its mostly choice and style I agree, but again !ptr is just very very common.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
-		fb[i + 1] = 0xff;
-		fb[i + 2] = 0x00;
+	touchmask = (uint8_t *)calloc(disp->fb_len, sizeof(uint8_t));
+	if (!touchmask)
```
see previous remark

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+static int renderloop(struct libevdev *evdev, struct display_info *disp,
+		     uint32_t xsize, uint32_t ysize, uint32_t fade, const bool banding)
+{
+	clock_t offset;
```
Declare and assign where you start using it, on line 372

oliv3r replied:
while that's not possible, the other away around could have been:
clock_t offset = clock();

While I don't think the initial nanoseconds difference wil matter, I initialize the clock just before starting the while loop, to be as accurate as possible.

raysiudak replied:
yes yes .. clock_t offset = clock();

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+
+	offset = clock();
+
+	while (true) {
```
Maybe a bit overkill, but good programming nonetheless.. install a signal handler to do a proper cleanup on exit.

oliv3r replied:
while ctrl-c takes care of all that implicitly, I agree it's better and more maintainable to do so implicitly. Added.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		int next_event;
+		int x, y;
+		struct input_event event;
+		uint32_t msec;
```
assign directly where you start using it.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+{
+	clock_t offset;
+	uint8_t *backbuffer, *touchmask;
+	uint32_t elapsed = 0;
```
Declare variables as close to the place where you start using them, in this case just above the while loop.

oliv3r replied:
actually, declare variables as cl the scope where you need them. I've added a fixup for that.

as for elapsed, it can't be closer, as it needs to be initialized with 0. The alternative is to declare it as a static inside the while loop which I find to be less accurate. as that variable will also hold its value outside of the renderloop function itself.

raysiudak replied:
Yes it should be just before the while loop, there is a lot of code in between

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	offset = clock();
+
+	while (true) {
+		int next_event;
```
see above comment, declare as close as possible to where you start using it

oliv3r replied:
she won't get any closer capt'n or we'll overload the fluxcapacitor!

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+			update_input = true;
+		}
+
+		if (!(msec % FPS(DISPLAY_FRAME_RATE))) {
```
'if not msec modulo FPS display frame rate', I prefer using != 0 when the thing being checked is zero or non zero.  
I prefer it the other way in case of line 392 if (!frame_drawn), reading 'if not frame drawn'

oliv3r replied:
yeah i forgot to change this

the main reason being, frame_drawn is an actual boolean, msec % module is an integer counter. I tend to get overeager in the !<int> usage, which I should avoid on integers.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+static struct libevdev *evdev_open(const char *path)
 {
-	int i;
+	struct libevdev *evdev = NULL;
```
Aha you are using NULL here :P

raysiudak replied:
And again declare as close as possible to first usage, i hate scrolling up :P

oliv3r replied:
typo!

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
 	}
+
+	ret = libevdev_new_from_fd(fd, &evdev);
+	if (ret) {
```
Returns negative when something goes wrong.
if (ret < 0)

oliv3r replied:
actually, it returns 0 on success, an error otherwise (either positive or negative) albeit granted, errors should and are nearly always negative.

that said, the reference docs also say to only check for negative, so i'll change that to match the [reference docs](https://www.freedesktop.org/software/libevdev/doc/0.2/group__init.html).

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static int is_event_device(const struct dirent *dir)
 {
+	return (strncmp(EVENT_DEV_NAME, dir->d_name, sizeof(EVENT_DEV_NAME) - 1) == 0);
```
What if dir is not initialized?

oliv3r replied:
[scandir guarantee's that dir is initialized, as it will call this function only if (dir)](https://github.com/walac/glibc/blob/master/dirent/scandirat.c#L80).

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
-		fb[i + 2] = 0xff;
+	fd = open(path, O_RDONLY | O_NONBLOCK);
+	if (fd < 0) {
+		fprintf(stderr, "Unable to open '%s': %m\n", path);
```
%m’ conversion, are we married to GNU C?

oliv3r replied:
lol, it's a GNU gcc extension; As I don't forse us ever using IAR or MSVC as compiler for our GNU/Linux applications.
I'll gladly change it to
```c
%s", strerror (errno)
```
If you really want it :)

But yes, I think we are actually more or less married to GNU in our current system :p as switching to a non-gnu system is almost impossible!

raysiudak replied:
No its fine, yes we are married to GNU likilly !

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static struct libevdev *evdev_get_device(char *path)
+{
+	struct libevdev *evdev;
```
Same comment here for local function members as give in previous comments

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		evdev = evdev_open(path);
 	}
+	if ((!evdev) || (!path))
+		goto err_out;
```
goto is soooo assembly.... This function is kind of messy with respect to failed branches and cleaning up, i think it should be rewritten with the use of some helpers maybe.

oliv3r replied:
goto? assembly? It is best avoided but there ARE valid usecases, exactly as you mention it, for cleanup.
```sh
~/src/jedi/um-kernel/linux $ git grep goto | wc -l
149325
~/src/jedi/um-u-boot/u-boot $ git grep goto | wc -l
3294
~/src/busybox]% git grep goto | wc -l
1798
```
So yes, you can put all the error cases in helper functions, but it tends to get even more messy and complex.

in this applications case, we'd only need a few helpers, 5 or 10 single purpose helpers.
But if you really want it, we can added it if you think its worth the time.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static int is_fb_device(const struct dirent *dir)
+{
+	return (strncmp(FB_DEV_NAME, dir->d_name, sizeof(FB_DEV_NAME) - 1) == 0);
```
what if dir is NULL

oliv3r replied:
[scandir guarantee's that dir is initialized, as it will call this function only if (dir)](https://github.com/walac/glibc/blob/master/dirent/scandirat.c#L80).

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static struct display_info *disp_open(const char *path)
+{
+	struct fb_var_screeninfo var_info;
```
Same here declare, where used

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	disp->fb_dev = open(path, O_RDWR);
+	if (disp->fb_dev < 0) {
+		fprintf(stderr, "Failed to open '%s': %s.\n", path, strerror(disp->fb_dev));
+		goto err_mem;
```
just free and return here no need for goto

oliv3r replied:
I like the central exit point pattern, where the only exception is if you can immediately return without any consequences.

I can be convinced to change the other exists to go to the same exit point, e.g. return NULL -> goto exit point at the end

raysiudak replied:
Yeah i agree that and i kind of come to appreciate the notation ( and you should keep it). Still it only saves 1 line in C, i am wondering about assembler (too lazy to test it :P)

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
-		fprintf(stderr, "Unable to get screen info: %s.\n", strerror(ret));
-
+		fprintf(stderr, "Unable to get var info: %s.\n", strerror(ret));
 		goto err_fb_dev;
```
same here

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
 	if (ret < 0) {
 		fprintf(stderr, "Unable to get fixed info: %s.\n", strerror(ret));
-
 		goto err_fb_dev;
```
and here :P, ok for the rest of the function...

raysiudak commented on src/ucit.c:
```diff
@@ -6,9 +6,17 @@
  *
```
You are using tabs instead of spaces.

oliv3r replied:
In the whole entire file! It is an outrage

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	int ret;
+
+	if (!path) {
+		ndev = scandir(DEV_FB, &namelist, is_fb_device, versionsort);
```
Memory leak: namelist is never freed

raysiudak replied:
ok more reasons to marry GNU C

oliv3r replied:
oh shit, yep nice catch, i completly missed that.

oliv3r replied:
I don't even wanna know the secret GNU stuff we use without knowing! Poor GNU C, nobody wants to marry it.

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+		disp = disp_open(path);
+	}
+	if (!disp || !path)
+		goto err_out;
```
goto is sooo assembly, i think this function can be split up and written more nicely to avoid this.

oliv3r replied:
this is nicely!

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static struct display_info *disp_get_device(char *path)
+{
+	struct display_info *disp;
```
same story on the local variables

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+ */
+static int parse_opts(int argc, char *argv[], char **fbpath, char **evpath, uint32_t *xsize, uint32_t *ysize, uint32_t *fadespeed, bool *banding)
+{
+	int c, option_index = 0;
```
one line for one variable

oliv3r replied:
however these variables are very closely related, they are both for get_opts. But to avoid the argument that they are not closley related but just a little bit, i'll split them :p

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+
+	*evpath = NULL;
+	*fbpath = NULL;
+	*banding = false;
```
shouldn't the rest of the arguments be initialized to 0 too?

oliv3r replied:
i was debating where they should be initialized. main() is the one doing the init now. However for the two pointers we get compiler errors about them being uninialized, so for some reason, when passing ** to a function, gcc assumes that's where they need to be initialized.

So either keep all init in main, or move all init here. having it here, can make sense, as this is where all options are parsed (and setup).

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+			*fbpath = strdup(optarg);
+			break;
+		case 't':
+			if (sscanf(optarg, "%ux%u", xsize, ysize) != 2) {
```
Why should you want to only specify the x size, i rarely see any square screens, i would prefer an explicit x and y option

oliv3r replied:
it does both actually :) (x[xy])

e.g. you supply -t 15 (for square) or -t 10x20 (for rectangular). it's simple and elegant!

raysiudak replied:
aha ok

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+
+int main(int argc, char *argv[])
+{
+	char *fbpath, *evpath;
```
one line for one variable and init to NULL

raysiudak commented on src/ucit.c:
```diff
@@ -17,142 +25,818 @@
+	evdev = evdev_get_device(evpath);
+	if (!evdev) {
+		ret = EXIT_FAILURE;
+		goto err_disp;
```
i think a goto is not really justified here

raysiudak replied:
Ok i see clearly now, it separates the error handling nicely, keep it.

oliv3r committed:
ci: Chicken and the egg.

When doing dockerimage changes we want to build the package using this
build environment. Two things are needed to accomplish this.

First; rules in the CI control file to use the appropiate image.

Secondly; the runner must have the pull-policy set to 'if-not-present'.

The latter however requires that we cleanup the server registry
occasionally of remote images.

Note that until we have sticky runners / workspace (gitlab-org/gitlab-ce#29447)
this is the best we can do.

Signed-off-by: Olliver Schinagl <oliver@schinagl.nl>

raysiudak reviewed (changes requested):
Not going to review the rewrite in the same pull request. This will make it really difficult to follow, make a new task or story for this and create a pull request for that. 

As production i want to twiddle parameters, so that ...

raysiudak commented on build.sh:
```diff
@@ -6,32 +6,94 @@
+    cmake \
+        -DCMAKE_C_COMPILER="${CROSS_COMPILE}gcc" \
+        -DARCH="${ARCH}" \
+        -H"." \
```
shouldn't there be spaces between argument and value

oliv3r replied:
according to the documentaion and examples no, they don't even quote the strings, but i found -H. to be very subtle to read. I think it even refused to work with the space! but i'm happy to retry now that it works (as it was not working before)
